src/components/application_manager/rpc_plugins/sdl_rpc_plugin/src/waypoints_pending_resumption_handler.cc

```diff
@@ -36,99 +36,104 @@ namespace sdl_rpc_plugin {
     : PendingResumptionHandler(application_manager) {}
 
 smart_objects::SmartObjectSPtr
 WayPointsPendingResumptionHandler::CreateSubscriptionRequest() {
   SDL_LOG_AUTO_TRACE();
+
   auto subscribe_waypoints_msg =
       application_manager::MessageHelper::CreateMessageForHMI(
-          hmi_apis::FunctionID::Navigation_SubscribeWayPoints,
+          Navigation_SubscribeWayPoints,
           application_manager_.GetNextHMICorrelationID());
   (*subscribe_waypoints_msg)[application_manager::strings::params]
                             [application_manager::strings::message_type] =
                                 hmi_apis::messageType::request;
   return subscribe_waypoints_msg;
 }
 
+void WayPointsPendingResumptionHandler::SendHMIRequestForNotSubscribed(
+    PendingRequest& pending_request) {
+  SDL_LOG_AUTO_TRACE();
+  SDL_LOG_DEBUG("Sending request with function id: "
+                << Navigation_SubscribeWayPoints
+                << " and correlation_id: " << pending_request.corr_id_);
+
+  auto request = application_manager::MessageHelper::CreateMessageForHMI(
+      Navigation_SubscribeWayPoints, pending_request.corr_id_);
+  (*request)[application_manager::strings::params]
+            [application_manager::strings::message_type] =
+                hmi_apis::messageType::request;
+  subscribe_on_event(Navigation_SubscribeWayPoints, pending_request.corr_id_);
+  application_manager_.GetRPCService().ManageHMICommand(request);
+  pending_request.waiting_for_hmi_response_ = true;
+}
+
 void WayPointsPendingResumptionHandler::HandleResumptionSubscriptionRequest(
     application_manager::AppExtension& extension,
     application_manager::Application& app) {
   SDL_LOG_AUTO_TRACE();
-  WayPointsAppExtension& ext = dynamic_cast<WayPointsAppExtension&>(extension);
-  smart_objects::SmartObjectSPtr request = CreateSubscriptionRequest();
-  smart_objects::SmartObject& request_ref = *request;
-  const auto function_id = static_cast<hmi_apis::FunctionID::eType>(
-      request_ref[application_manager::strings::params]
-                 [application_manager::strings::function_id]
-                     .asInt());
-  const uint32_t corr_id =
-      request_ref[application_manager::strings::params]
-                 [application_manager::strings::correlation_id]
-                     .asUInt();
+  sync_primitives::AutoLock lock(pending_resumption_lock_);
+  UNUSED(extension);
+
+  if (application_manager_.IsAnyAppSubscribedForWayPoints()) {
+    SDL_LOG_DEBUG(
+        "Subscription to waypoint already exist, no need to send "
+        "request to HMI");
+    application_manager_.SubscribeAppForWayPoints(app.app_id());
+    return;
+  }
 
+  const auto request = CreateSubscriptionRequest();
+  const auto corr_id = get_corr_id_from_message(*request);
   auto resumption_request =
-      MakeResumptionRequest(corr_id, function_id, *request);
-  app_ids_.push(app.app_id());
+      MakeResumptionRequest(corr_id, Navigation_SubscribeWayPoints, *request);
```

Comment:
This is an internal function?. 
Suppose that  `WayPointsPendingResumptionHandler::MakeResumptionRequest` works only with `Navigation_SubscribeWayPoints`. So this parameter in the function seems redundant.

Reply:
This function is a method, inherited from parent class PendingResumptionHandler. So it works with any type of request.